main.tf

```diff
@@ -1,25 +1,38 @@
 data "aws_caller_identity" "current" {
 }
 
+data "aws_region" "current" {
+}
+
 locals {
-  account_id = data.aws_caller_identity.current.account_id
+  account_id  = data.aws_caller_identity.current.account_id
+  bucket_name = var.bucket_name == null ? "${local.account_id}-${local.region}-s3logging-${var.bucket_suffix}" : var.bucket_name
+  region      = data.aws_region.current.name
 }
 
-resource "aws_s3_bucket" "s3logging_bucket" {
-  bucket = "${local.account_id}-${var.region}-s3logging-${var.bucket_suffix}"
+# Ignore logging requirement - access logging for a logging bucket is a little meta
+#tfsec:ignore:AWS002
+resource "aws_s3_bucket" "this" {
+  bucket = local.bucket_name
   acl    = "log-delivery-write"
   tags   = var.tags
 
-  versioning {
-    enabled = true
-  }
+  dynamic "lifecycle_rule" {
+    iterator = rule
+    for_each = var.lifecycle_rules
+
+    content {
+      id      = rule.value.id
+      enabled = rule.value.enabled
+      prefix  = lookup(rule.value, "prefix", null)
 
-  lifecycle_rule {
-    id      = "expire"
-    enabled = true
+      expiration {
+        days = lookup(rule.value, "expiration", 2147483647)
+      }
 
-    noncurrent_version_expiration {
-      days = 90
+      noncurrent_version_expiration {
+        days = lookup(rule.value, "noncurrent_version_expiration", 365)
+      }
     }
```

Comment:
It looks like we're keeping logs (essentially) forever but only keeping proof that they aren't being tampered with for a year. Is that the desired outcome?

Also instead of using an absurdly long default expiration, could you use a `try()` statement to only set an expiration if an expiration rule is passed?

Reply:
That's just the max expiration that S3 allows, and unless I'm not seeing the path to it, you couldn't do try on a block. Setting null isn't supported for this value.

Reply:
I was trying to suggest using `try()` (or similar) to exclude the entire block if `null`

Reply:
To my knowledge, you can't use try() to exclude a _block_. More importantly, the only thing this thing does is define expiration for the files (and the versions - that was intended to be the same value for obvious reasons). So, if it's all empty, why have it at all?

Reply:
That's a good point.

I think I'd prefer to see a validation rule requiring a non-null value within whatever valid range, but this is fine for now.

If you _do_ pass a `null`, does it fail at plan time or apply time? If it fails at plan time it might be fine to just not have a default and let the provider fail it.